File: hash/rust/rosetta-image-hash/tests/group1_gaussian_blur_test.rs
mod testkit;

use rosetta_image_hash::internal::pil_gaussian_blur;
use serde::Deserialize;
use std::fs;
use testkit::spec_path::SPEC_DIR;

#[derive(Debug, Deserialize)]
struct GaussianBlurCase {
    name: String,
    input: Vec<Vec<u8>>,
    output: Vec<Vec<u8>>,
}

#[derive(Debug, Deserialize)]
struct GaussianBlurCases {
    cases: Vec<GaussianBlurCase>,
}

fn load_cases() -> GaussianBlurCases {
    let path = format!("{}/gaussian_blur_cases.json", SPEC_DIR);
    let data = fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {path}: {e}"));
    serde_json::from_str(&data).expect("parse gaussian_blur_cases.json")
}

#[test]
fn gaussian_blur_group1() {
    let cases = load_cases();
    let mut failures: Vec<String> = Vec::new();

    for c in &cases.cases {
        let got = pil_gaussian_blur::apply(&c.input, 2.0);

        if got != c.output {
            // Find first mismatch
            let mut first_diff = String::new();
            'outer: for y in 0..c.output.len() {
                for x in 0..c.output[y].len() {
                    if got[y][x] != c.output[y][x] {
                        first_diff = format!(
                            "first diff at ({},{}) got={} want={}",
                            y, x, got[y][x], c.output[y][x]
                        );
                        break 'outer;
                    }
                }
            }
            failures.push(format!("case={} {}", c.name, first_diff));
        }
    }

    if !failures.is_empty() {
        panic!("{} gaussian_blur failures:\n  {}", failures.len(), failures.join("\n  "));
    }
}

File: hash/js/rosetta-squint-hash/tests/group1-gaussian-blur.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { SPEC_DIR } from "./testkit.js";
import { pilGaussianBlur } from "../src/internal/pilGaussianBlur.js";

interface GaussianBlurCase {
  name: string;
  shape: [number, number];
  input: number[][];
  output: number[][];
}

interface GaussianBlurCases {
  cases: GaussianBlurCase[];
}

describe("pilGaussianBlur", () => {
  it("byte-exact against all 6 gaussian_blur_cases.json cases", () => {
    const path = join(SPEC_DIR, "gaussian_blur_cases.json");
    const doc = JSON.parse(readFileSync(path, "utf8")) as GaussianBlurCases;
    expect(doc.cases.length).toBe(6);

    const failures: string[] = [];
    for (const c of doc.cases) {
      const [height, width] = c.shape;
      const flat = new Uint8Array(height * width);
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          flat[y * width + x] = c.input[y][x];
        }
      }
      const got = pilGaussianBlur(flat, width, height);
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          const idx = y * width + x;
          if (got[idx] !== c.output[y][x]) {
            failures.push(
              `${c.name} (${y},${x}): got ${got[idx]} want ${c.output[y][x]}`
            );
          }
        }
      }
    }
    if (failures.length > 0) {
      throw new Error(`${failures.length} pixel failures:\n  ${failures.slice(0, 10).join("\n  ")}`);
    }
  });
});
